client/rpc/manual-xcm/src/lib.rs

```diff
@@ -0,0 +1,126 @@
+
+/// This RPC interface is used to manually submit XCM messages that will be injected into a
+/// parachain-enabled runtime. This allows testing XCM logic in a controlled way in integration
+/// tests.
+#[rpc(server)]
+pub trait ManualXcmApi {
+	/// Inject a downward xcm message - A message that comes from the relay chain.
+	/// You may provide an arbitrary message, or if you provide an emtpy byte array,
+	/// Then a default message (DOT transfer down to ALITH) will be injected
+	#[rpc(name = "xcm_injectDownwardMessage")]
+	fn inject_downward_message(&self, message: Vec<u8>) -> BoxFuture<'static, RpcResult<()>>;
+
+	/// Inject an HRMP message - A message that comes from a dedicated channel to a sibling
+	/// parachain.
+	///
+	/// Cumulus Parachain System seems to have a constraint that at most one hrmp message will be
+	/// sent on a channel per block. At least that's what this comment implies:
+	/// https://github.com/paritytech/cumulus/blob/c308c01b/pallets/parachain-system/src/lib.rs#L204
+	/// Neither this RPC, nor the mock inherent data provider make any attempt to enforce this
+	/// constraint. In fact, violating it may be useful for testing.
+	#[rpc(name = "xcm_injectHrmpMessage")]
+	fn inject_hrmp_message(
+		&self,
+		sender: ParaId,
+		message: Vec<u8>,
+	) -> BoxFuture<'static, RpcResult<()>>;
+}
+
+pub struct ManualXcm {
+	pub downward_message_channel: flume::Sender<Vec<u8>>,
+	pub hrmp_message_channel: flume::Sender<(ParaId, Vec<u8>)>,
+}
+
+impl ManualXcmApi for ManualXcm {
+	fn inject_downward_message(&self, msg: Vec<u8>) -> BoxFuture<'static, RpcResult<()>> {
+		let downward_message_channel = self.downward_message_channel.clone();
+		async move {
+			// If no message is supplied, inject a default one.
+			let msg = if msg.is_empty() {
+				xcm::VersionedXcm::<()>::V2(Xcm(vec![
+					ReserveAssetDeposited((Parent, 10000000000000).into()),
+					ClearOrigin,
+					BuyExecution {
+						fees: (Parent, 10000000000000).into(),
+						weight_limit: Limited(4_000_000_000),
+					},
+					DepositAsset {
+						assets: All.into(),
+						max_assets: 1,
+						beneficiary: MultiLocation::new(
+							0,
+							X1(AccountKey20 {
+								network: Any,
+								key: hex_literal::hex!("f24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"),
+							}),
+						),
+					},
+				]))
+				.encode()
+			} else {
+				msg
+			};
+
+			// Push the message to the shared channel where it will be queued up
+			// to be injected in to an upcoming block.
+			downward_message_channel
+				.send_async(msg)
+				.await
+				.map_err(|err| internal_err(err))?;
+
+			Ok(())
+		}
+		.boxed()
+	}
+
+	fn inject_hrmp_message(
+		&self,
+		sender: ParaId,
+		msg: Vec<u8>,
+	) -> BoxFuture<'static, RpcResult<()>> {
+		let hrmp_message_channel = self.hrmp_message_channel.clone();
+
+		async move {
+			// Push the message and sender to the shared channel where they will be queued up
+			// to be injected in to an upcoming block.
+			hrmp_message_channel
```

Comment:
I feel we can do the same as before here. The message would change a bit in this case, being

```
xcm::VersionedXcm::<()>::V2(Xcm(vec![
					ReserveAssetDeposited(((Parent, Parachain(sender)).into(), 10000000000000).into()),
					ClearOrigin,
					BuyExecution {
						fees: (Parent, Parachain(sender)).into(),
						weight_limit: Limited(4_000_000_000),
					},
					DepositAsset {
						assets: All.into(),
						max_assets: 1,
						beneficiary: MultiLocation::new(
							0,
							X1(AccountKey20 {
								network: Any,
								key: hex_literal::hex!("f24FF3a9CF04c71Dbc94D0b566f7A27B94566cac"),
							}),
						),
					},
				]))
				.encode()
```

Reply:
This is a good idea, but I couldn't quite get it to compile. I left the code in but commented with a todo. We can handle that here or in a followup.